### PR TITLE
crimson,mon: add guards to avoid accidental crimson deployment and to avoid usage of unsupported features with crimson

### PR DESCRIPTION
**Pull request by athanatos**
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
crimson-rados run: https://pulpito.ceph.com/sjust-2022-10-17_00:27:00-crimson-rados-wip-sjust-testing-2022-10-16-150616-distro-default-smithi/

Also need a rados run due to monitor changes: https://pulpito.ceph.com/sjust-2022-10-17_17:00:50-rados-wip-sjust-testing-2022-10-16-150616-distro-default-smithi/

**Comment by athanatos** [comment]
jenkins test make check

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -3156,7 +3168,7 @@ void OSDMap::encode(ceph::buffer::list& bl, uint64_t features) const
   {
     // NOTE: any new encoding dependencies must be reflected by
     // SIGNIFICANT_FEATURES
-    uint8_t target_v = 9; // when bumping this, be aware of range blocklist
+    uint8_t target_v = 9; // when bumping this, be aware of allow_crimson
```
Comment (Matan-B) [inline src/osd/OSDMap.cc:3171]: Shouldn't we set `target_v` to 12? Same for `DECODE_START`? (e.g 2be95af8211225c43d72e439b6000a25743bdab7)

Comment (athanatos) [reply src/osd/OSDMap.cc:3171]: No.  The idea here is that we avoid setting the more restrictive version if allow_crimson isn't set.  Just below, if allow_crimson is set, we bump target_v to 12.

**Comment by athanatos** [comment]
Another crimson run, teuthology changes seem to be working ok, failures seem unrelated: https://pulpito.ceph.com/sjust-2022-12-12_07:11:41-crimson-rados-wip-sjust-testing-2022-12-10-142339-distro-default-smithi/

**Comment by athanatos** [comment]
@rzarzynski @gregsfortytwo I'd like to merge this asap, do you have time to review it?

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -2488,6 +2498,8 @@ int OSDMap::apply_incremental(const Incremental &inc)
     stretch_mode_bucket = inc.new_stretch_mode_bucket;
   }
 
+  allow_crimson ^= inc.toggle_allow_crimson;
```
Comment (rzarzynski) [inline src/osd/OSDMap.cc:2501]: xor should be fine assuming we really want the _toggle_ semantic and the event's life-time is one epoch only.

File: src/mon/MonCommands.h [hunk 3]
```diff
@@ -1164,6 +1164,13 @@ COMMAND("osd force_recovery_stretch_mode " \
 	"pool size to its non-failure value if currently degraded and "
 	"all monitor buckets are up",
 	"osd", "rw")
+COMMAND("osd set-allow-crimson " \
+	"name=yes_i_really_mean_it,type=CephBool,req=false",
+	"Allow crimson-osds to boot and join the cluster.  Note, crimson-osd is "
+	"not yet considered stable and may crash or cause data loss -- should "
+	"be avoided outside of testing and development.  This setting is "
+	"irrevocable",
```
Comment (rzarzynski) [inline src/mon/MonCommands.h:1173]: `or` instead `xor` in the incremental map handling? Do we have a plan to change this claim in the future?

File: src/mon/OSDMonitor.cc [hunk 4]
```diff
@@ -3552,6 +3552,21 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
     goto ignore;
   }
 
+  // See crimson/osd/osd.cc: OSD::_send_boot
+  if (m->metadata.contains("osd_type")) {
+    // m->metadata["osd_type"] must be "crimson", classic doesn't send osd_type
```
Comment (rzarzynski) [inline src/mon/OSDMonitor.cc:3557]: assert?

File: src/mon/OSDMonitor.cc [hunk 5]
```diff
@@ -8419,6 +8419,10 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
     }
     p.min_size = n;
   } else if (var == "pg_num_actual") {
+    if (p.has_flag(pg_pool_t::FLAG_NOPGCHANGE)) {
```
Comment (rzarzynski) [inline src/mon/OSDMonitor.cc:8424]: This commit seems to be unrelated to crimson. At first glance looks sane but maybe could be sent separately to facilitate broader review.

File: src/mon/OSDMonitor.cc [hunk 6]
```diff
@@ -8169,8 +8169,10 @@ int OSDMonitor::prepare_new_pool(string& name,
     pi->use_gmt_hitset = true;
   else
     pi->use_gmt_hitset = false;
-  if (crimson)
+  if (crimson) {
     pi->set_flag(pg_pool_t::FLAG_CRIMSON);
+    pi->set_flag(pg_pool_t::FLAG_NOPGCHANGE);
```
Comment (rzarzynski) [inline src/mon/OSDMonitor.cc:8174]: OK, I see.

File: src/osd/OSDMap.cc (hunk 2 above)
Comment (athanatos) [reply src/osd/OSDMap.cc:2501]: The "event" is the toggle.  We will eventually want to be able to turn allow_crimson off.

File: src/mon/MonCommands.h (hunk 3 above)
Comment (athanatos) [reply src/mon/MonCommands.h:1173]: It's irrevocable for now as a matter of policy, but we'll probably relax that restraint in the future.

File: src/mon/OSDMonitor.cc (hunk 4 above)
Comment (athanatos) [reply src/mon/OSDMonitor.cc:3557]: No, I don't want to crash the mon.

File: src/osd/OSDMap.cc (hunk 2 above)
Comment (rzarzynski) [reply src/osd/OSDMap.cc:2501]: OK, I see – low-level for the future.

File: src/mon/OSDMonitor.cc (hunk 6 above)
Comment (athanatos) [reply src/mon/OSDMonitor.cc:8174]: Yeah, it's there because crimson can't do splits or merges yet.

**Review by gregsfortytwo — CHANGES_REQUESTED** [review]
I think this pokes at the right places; I just have some stronger qualms over some of the bits @rzarzynski pointed out I'd like to adjust. :)

File: src/mon/OSDMonitor.cc (hunk 4 above)
Comment (athanatos) [reply src/mon/OSDMonitor.cc:3559]: Fair point, fixed to output a message to derr but to otherwise allow it.

File: src/osd/OSDMap.cc (hunk 2 above)
Comment (athanatos) [reply src/osd/OSDMap.cc:2501]: I suppose I could make it an enum.

Comment (athanatos) [reply src/osd/OSDMap.cc]: Ok, changed it to an enum.  Preserved the ability for the Incremental application to remove the flag in case we want to backport that kind of capability later on.

**Comment by athanatos** [comment]
Note: need to rerun tests

File: src/osd/OSDMap.cc (hunk 2 above)
Comment (gregsfortytwo) [reply src/osd/OSDMap.cc]: The field could be named "crimson_enabled", and just set the osdmap flag equal to it? That does mean you can't identify if a given incremental causes a change without the preceding osdmap, but is that actually a problem?

Comment (gregsfortytwo) [reply src/osd/OSDMap.cc]: ^ I guess because that makes it hard to conditionally encode the incremental with the different struct_v?

Anyway, the enum version seems fine.

**Review by gregsfortytwo — APPROVED** [review]
LGTM

**Comment by athanatos** [comment]
jenkins retest this please

**Comment by athanatos** [comment]
crimson-rados run looks good: https://pulpito.ceph.com/sjust-2023-01-30_19:56:26-crimson-rados-wip-sjust-58486-distro-default-smithi/

**Comment by athanatos** [comment]
crimson-rados: https://pulpito.ceph.com/sjust-2023-02-14_21:48:25-crimson-rados-wip-sjust-57801-distro-default-smithi/ failures are the usual from main
 
rados: still queued https://pulpito.ceph.com/sjust-2023-02-14_21:12:04-rados-wip-sjust-57801-distro-default-smithi

**Comment by athanatos** [comment]
Added needs-qa flag since my run seems like it may be stuck in the queue.

**Comment by athanatos** [comment]
Canceled the stuck one, trying to rerun https://pulpito.ceph.com/sjust-2023-02-16_21:39:46-rados-wip-sjust-57801-distro-default-smithi/

**Comment by athanatos** [comment]
@yuriw @ljflores Sorry to bug you, but would it be possible to get this in your next round of integration testing?  Rados runs I've been scheduling don't seem to be running presumably because the lab is oversubscribed, but it's important that this get tested/merged prior to the reef freeze.

**Comment by athanatos** [comment]
https://pulpito.ceph.com/yuriw-2023-02-22_00:06:40-rados-wip-yuri6-testing-2023-02-21-1406-distro-default-smithi/ Failures seem unrelated!
